topology_custom/test_tablets: add remove/replace tests for edge cases

bhalevy opened the pull request:
Test cases related to #21826:

1. test_remove_failure_with_no_normal_token_owners_in_dc: attempts to
    remove a node with another node down in the datacenter, leaving
    no normal token owners in that dc (reproducing #21826).
    Removenode is expected to fail in this case since it
    should have no place to rebuild the removed node replicas,
    yet it currently succeeds unexpectedly.

2. test_remove_failure_then_replace: verify that removenode
    fails as expected when there are not enough nodes to
    rebuild its replicas on, with and without additional zero-token nodes.

3. test_replace_with_no_normal_token_owners_in_dc: verify that
    nodes can be replaced in a datacenter that has no live
    token owners, with and without additional zero-token nodes.
    Tablet replace uses all replicas to rebuild the lost replicas
    and therefore should succeed in the edge case.
    The restored data is verified as well.

Refs #21826

* New tests, no backport needed

bhalevy commented:
In v2 ([6e96cdd](https://github.com/scylladb/scylladb/commit/6e96cdd8544b6b5ea432b153ddb14105cf6def9a)):
- test: pylib: _cluster_remove_node: mark server as removed only when removenode succeeded
- test: pylib: _cluster_remove_node: log message on successful paths
- test: topology_custom: add test_remove_failure_then_replace_with_zero_token_node
  - verify replace after failed removenode

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13669/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13669/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_tablets
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/557/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13669/)
#### Build Details:
- Duration: 5 hr 46 min
- Builder: spider1.cloudius-systems.com

bhalevy commented:
rebased

bhalevy commented:
@kbr-scylla can you please review too?

bhalevy commented on test/topology_custom/test_tablets.py:
```diff
@@ -626,3 +627,35 @@ async def test_orphaned_sstables_on_startup(manager: ManagerClient):
+    logger.info("Attempting removenode - expected to fail")
+    await manager.server_stop_gracefully(node_to_remove.server_id)
+    await manager.remove_node(initiator_node.server_id, server_id=node_to_remove.server_id,
+                              expected_error="Removenode failed. See earlier errors (Rolled back: Failed to drain tablets: std::runtime_error (Unable to find new replica for tablet")
```
Done in next version

bhalevy commented:
In v3 ([e591d67](https://github.com/scylladb/scylladb/commit/e591d67e36c86e3ac1b7e15b316d5cee7cb001fa)):
- changed initial topology to have an additional zero-token node in dc1
- changed rf to `{'dc1': 2, 'dc2': 1}` although tablets should never migrate between DCs, let's strengthen the tests in this area.

bhalevy commented on test/topology_custom/test_tablets.py:
```diff
@@ -626,3 +627,34 @@ async def test_orphaned_sstables_on_startup(manager: ManagerClient):
+    await cql.run_async(f"CREATE KEYSPACE test WITH replication = {{ 'class': 'NetworkTopologyStrategy', 'dc1': 2, 'dc2': 1 }} AND tablets = {{ 'initial': 1 }}")
+    await cql.run_async("CREATE TABLE test.test (pk int PRIMARY KEY, c int);")
+    num_keys = 1000
+    await asyncio.gather(*[cql.run_async(f"INSERT INTO test.test (pk, c) VALUES ({k}, {k%3});") for k in range(num_keys)])
```
dropped in v3

bhalevy commented:
In v3 (b953120):
- Added more test cases
- In particular, reproducer for #21826 (currently marked as xfail)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13748/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13748/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_tablets
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/618/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13748/)
#### Build Details:
- Duration: 7 hr 3 min
- Builder: spider7.cloudius-systems.com
